### PR TITLE
Reject workspace creation if namespace already exists

### PR DESCRIPTION
**Pull request by raulcabello**
## Issue: 
https://github.com/rancher/fleet/issues/1519
 
## Problem
Rancher allows the creation of fleet workspaces even though the namespace already exists. Deleting that workspace causes the deletion of namespace along with its contents. Deleting unwanted fleet workspace results in data loss.

## Solution
Reject the workspace creation if namespace already exists

### CODE REVIEW

**Review by weyfonk — APPROVED** [review]

File: pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go [hunk 1]
```diff
@@ -89,21 +91,18 @@ func (m *Mutator) Admit(request *admission.Request) (*admissionv1.AdmissionRespo
...
-		if err != nil {
-			return nil, err
-		}
-		response := &admissionv1.AdmissionResponse{}
-		auth.SetEscalationResponse(response, auth.ConfirmNoEscalation(request, cr.Rules, namespace.Name, m.resolver))
-
-		ns, err = m.namespaces.Get(namespace.Name, metav1.GetOptions{})
-		if err != nil {
-			return nil, err
+		if errors.IsAlreadyExists(err) {
```
Comment (KevinJoiner) [inline pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go:94]: This will make the mutator no longer [Idempotent](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#idempotence). kube-apiserver may call mutators multiple times for one request, which would cause this to fail the second time it is ran [more info on reinvocation](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#reinvocation-policy)

One way to solve the original issue is to add labels on the created namespace that indicates it was created for fleet. Then you could deny the creation of a fleet workspace for any namespaces that do not have that label. And leave the current logic for a namespace that does have that label.

**Review by KevinJoiner — CHANGES_REQUESTED** [review]
This will also need unit tests for these new changes. You can check out some of the unit tests for clusterroletemplatebindings or secrets to get started.

File: pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go (hunk 1 above)
Comment (raulcabello) [reply pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go]: This has been fix in https://github.com/rancher/webhook/pull/260/commits/803504b098c5917b1bdff52c8ecc7f91148d7df8. I also added unit tests.

File: pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go [hunk 2]
```diff
@@ -19,6 +21,8 @@ import (
 	"k8s.io/kubernetes/pkg/registry/rbac/validation"
 )
 
+const fleetManagedLabel = "fleet-workspace-managed-resource"
```
Comment (KevinJoiner) [inline pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go:24]: I would suggest using a standard label like `app.kubernetes.io/managed-by: rancher`
https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/

File: pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go [hunk 3]
```diff
@@ -84,26 +88,31 @@ func (m *Mutator) Admit(request *admission.Request) (*admissionv1.AdmissionRespo
...
+		if ns.Labels[fleetManagedLabel] != "true" {
+			return &admissionv1.AdmissionResponse{
+				Result: &metav1.Status{
+					Status:  "Failure",
+					Message: fmt.Sprintf("namespace '%s' already exists", fw.Name),
+					Reason:  metav1.StatusReasonInvalid,
+					Code:    http.StatusBadRequest,
+				},
+				Allowed: false,
+			}, nil
```
Comment (KevinJoiner) [inline pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go:113]: use the admission helper function for consistent webhook messages.
```suggestion
			return admission.ResponseBadRequest(fmt.Sprintf("namespace '%s' already exists", fw.Name)), nil
```

File: pkg/resources/management.cattle.io/v3/fleetworkspace/validator_test.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+func TestAdmit(t *testing.T) {
+	nsName := "test"
+
+	nsExistMutator := func() Mutator {
+		ctrl := gomock.NewController(t)
+		mockNamespaceController := fakes.NewMockNamespaceController(ctrl)
```
Comment (KevinJoiner) [inline pkg/resources/management.cattle.io/v3/fleetworkspace/validator_test.go:24]: We no longer need to generate fakes for Wrangler controllers. Instead, we can use the new [generic/fake](https://github.com/rancher/wrangler/tree/master/pkg/generic/fake) package from wrangler.
Hear is an [example](https://github.com/rancher/webhook/pull/262/files#diff-7d2acc95d10395b594e7b75b2516089d511d173afa33235d3a8d0d87de41ea44) of how that change looks.

This uses gomock as well, so you shouldn't need to change any of your test logic.

**Review by KevinJoiner — CHANGES_REQUESTED** [review]
This approach looks like it should work. Before we merge, can you squash your commits into one single commit?

File: pkg/resources/management.cattle.io/v3/fleetworkspace/validator_test.go (hunk 4 above)
Comment (raulcabello) [reply pkg/resources/management.cattle.io/v3/fleetworkspace/validator_test.go]: done in https://github.com/rancher/webhook/pull/260/commits/24334874f8922d5187275b7fdf0f1c85dfcacf5f

File: pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go (hunk 3 above)
Comment (raulcabello) [reply pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go]: done in https://github.com/rancher/webhook/pull/260/commits/24334874f8922d5187275b7fdf0f1c85dfcacf5f

File: pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go (hunk 2 above)
Comment (raulcabello) [reply pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go]: done in https://github.com/rancher/webhook/pull/260/commits/24334874f8922d5187275b7fdf0f1c85dfcacf5f

**Comment by raulcabello** [comment]
@KevinJoiner I made the changes suggested, and I squashed the commits into a single commit.

File: pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go [hunk 5]
```diff
@@ -84,25 +87,21 @@ func (m *Mutator) Admit(request *admission.Request) (*admissionv1.AdmissionRespo
 
 	namespace := v1.Namespace{
 		ObjectMeta: metav1.ObjectMeta{
-			Name: fw.Name,
+			Name:   fw.Name,
+			Labels: map[string]string{k8sManagedLabel: "true"},
```
Comment (KevinJoiner) [inline pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go:91]: We should use the value of `rancher` here so users who see these namespaces know that Rancher is responsible for managing their creation and deletion.
```suggestion
			Labels: map[string]string{k8sManagedLabel: "rancher"},
```

File: pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go [hunk 6]
```diff
@@ -1,10 +1,11 @@
 package fleetworkspace
```
Comment (KevinJoiner) [inline pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go:1]: Can you rename this file and the test file to `mutator.go` and `mutator_test.go`? I think this was a mistake by someone else when the file was first created.

File: pkg/resources/management.cattle.io/v3/fleetworkspace/validator_test.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+		mockClusterRoleController := fake.NewMockNonNamespacedControllerInterface[*rbacv1.ClusterRole, *rbacv1.ClusterRoleList](ctrl)
+		mockClusterRoleController.EXPECT().Create(gomock.Any())
+
+		return Mutator{
+			namespaces:          mockNamespaceController,
+			rolebindings:        mockRoleBindingController,
+			clusterrolebindings: mockClusterRoleBindingController,
+			clusterroles:        mockClusterRoleController,
+		}
+	}
```
Comment (KevinJoiner) [inline pkg/resources/management.cattle.io/v3/fleetworkspace/validator_test.go:53]: These functions don't need to be local variables in the test function. They can be moved out into regular functions that take in `t *testing.T` as a parameter or `ctrl *gomock.Controler`

File: pkg/resources/management.cattle.io/v3/fleetworkspace/validator_test.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			m := test.m()
+			admit, _ := m.Admit(test.req)
+			if admit.Allowed != test.expectAllowed {
+				t.Errorf("Response was incorrectly validated, wanted response.Allowed = '%v' got %v: result=%+v", test.expectAllowed, admit.Allowed, admit.Result)
+			}
```
Comment (KevinJoiner) [inline pkg/resources/management.cattle.io/v3/fleetworkspace/validator_test.go:93]: We use testify for our condition checking which gives us uniformed error messages between our unit test.
```suggestion
			require.Equalf(test.expectAllowed, admit.Allowed, "Response was incorrectly validated, wanted response.Allowed = '%v' got %v: result=%+v", test.expectAllowed, admit.Allowed, admit.Result)
```

**Review by KevinJoiner — CHANGES_REQUESTED** [review]
We recently added a new requirement for documenting webhook functionality.
Could you also add a `FleetWorkspace.md` file that explains what this mutator is doing. 
Here is [how our docs work ](https://github.com/rancher/webhook#docs) and an [example Markdown file](https://github.com/rancher/webhook/blob/release/v0.4/pkg/resources/core/v1/secret/Secret.md) for our Secret mutator.

File: pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go (hunk 5 above)
Comment (raulcabello) [reply pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go]: done in https://github.com/rancher/webhook/pull/260/commits/bc3715088473c65ec63c87a6da489240e56dbaef

File: pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go (hunk 6 above)
Comment (raulcabello) [reply pkg/resources/management.cattle.io/v3/fleetworkspace/validator.go:1]: done in https://github.com/rancher/webhook/pull/260/commits/bc3715088473c65ec63c87a6da489240e56dbaef

File: pkg/resources/management.cattle.io/v3/fleetworkspace/validator_test.go (hunk 7 above)
Comment (raulcabello) [reply pkg/resources/management.cattle.io/v3/fleetworkspace/validator_test.go]: done in https://github.com/rancher/webhook/pull/260/commits/bc3715088473c65ec63c87a6da489240e56dbaef

File: pkg/resources/management.cattle.io/v3/fleetworkspace/validator_test.go (hunk 8 above)
Comment (raulcabello) [reply pkg/resources/management.cattle.io/v3/fleetworkspace/validator_test.go]: done in https://github.com/rancher/webhook/pull/260/commits/bc3715088473c65ec63c87a6da489240e56dbaef

**Comment by raulcabello** [comment]
@KevinJoiner I added the docs and made all the changes. Let me know what you think. Thanks!

File: pkg/resources/management.cattle.io/v3/fleetworkspace/mutator.go [hunk 9]
```diff
@@ -84,25 +88,29 @@ func (m *Mutator) Admit(request *admission.Request) (*admissionv1.AdmissionRespo
...
 	ns, err := m.namespaces.Create(&namespace)
 	if err != nil {
 		if !errors.IsAlreadyExists(err) {
 			return nil, err
 		}
-		// check if user has enough privilege to create fleet admin rolebinding in the namespace
-		cr, err := m.clusterroles.Cache().Get(fleetAdminRole)
+		ns, err := m.namespaces.Get(fw.Name, metav1.GetOptions{})
 		if err != nil {
 			return nil, err
```
Comment (KevinJoiner) [inline pkg/resources/management.cattle.io/v3/fleetworkspace/mutator.go:102]:
```suggestion
			return nil, fmt.Errorf("failed to get fleetworkspace namespace '%s': %w, fw.Name, err)
```

File: pkg/resources/management.cattle.io/v3/fleetworkspace/mutator_test.go [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+		},
+	}
+
+	tests := map[string]struct {
+		m                  func(t *testing.T) Mutator
+		req                *admission.Request
+		expectAllowed      bool
+		expectResultStatus *metav1.Status
+	}{
+		"reject because namespace already exists": {
```
Comment (KevinJoiner) [inline pkg/resources/management.cattle.io/v3/fleetworkspace/mutator_test.go:42]: Let's add the following test cases to hit all of the changes made in this PR.

- namespace already exists with the expected label
- namespace already exists with the expected label, but the user fails the escalation check
- failed to create a namespace with an error other than alreadyExist.

To add tests that hit the escalation check, you will need to create a resolver for your mutator there is an example of this [here](https://github.com/rancher/webhook/blob/release/v0.4/pkg/resources/management.cattle.io/v3/roletemplate/validator_test.go#L34C1-L50)

**Review by KevinJoiner — CHANGES_REQUESTED** [review]
Thanks for adding the docs. After you make changes, you can re-request a review in the top right corner, and I will get the notification.

File: pkg/resources/management.cattle.io/v3/fleetworkspace/mutator_test.go (hunk 10 above)
Comment (raulcabello) [reply pkg/resources/management.cattle.io/v3/fleetworkspace/mutator_test.go:53]: Done. Please check the latest commit.

**Review by KevinJoiner — DISMISSED** [review]
Thanks for making all the updates; this looks good to go.

File: pkg/resources/management.cattle.io/v3/fleetworkspace/mutator_test.go [hunk 11]
```diff
@@ -0,0 +1,263 @@
...
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			m := test.m(t)
+			admit, err := m.Admit(test.req)
+			if err != nil {
+				require.Equalf(t, test.expectedErr, err, "Error doesn't match, wanted %v got %v", test.expectedErr, err)
+				return
+			}
```
Comment (weyfonk) [inline pkg/resources/management.cattle.io/v3/fleetworkspace/mutator_test.go:102]: Making the check on the returned error, instead of the expected one, may expose us to unnoticed failures if no error is returned although one is expected.
Conversely, running equality checks only if the expected error is not `nil` means that if no error is expected but one is returned, it may fly under our radar.

TL;DR:
```suggestion
			if test.expectedErr == nil {
				require.NoError(t, err)
		        } else {
				require.Equalf(t, err, "Error doesn't match, wanted %v got %v", test.expectedErr, err)
			}
```

**Review by weyfonk — CHANGES_REQUESTED** [review]

File: pkg/resources/management.cattle.io/v3/fleetworkspace/mutator_test.go (hunk 11 above)
Comment (raulcabello) [reply pkg/resources/management.cattle.io/v3/fleetworkspace/mutator_test.go]: changed here https://github.com/rancher/webhook/pull/260/commits/fc284bd621528d5128f03b4573f4c60b05afc6dc
